### ISSUE TITLE
Considering the File System API

### ISSUE THREAD

**Issue by andreubotella** [wintercg/proposal-common-minimum-api#73]
The File System API is a relatively recent web standard, now implemented in all browsers, that allows accessing and modifying files in the file system. All browsers allow access through this API to a virtual filesystem, unique to a page's origin, through [`await navigator.storage.getDirectory()`](https://developer.mozilla.org/en-US/docs/Web/API/StorageManager/getDirectory), that you can't go outside of.

However, Chromium extends this by providing additional APIs (such as [`window.showOpenFilePicker()`](https://developer.mozilla.org/en-US/docs/Web/API/Window/showOpenFilePicker)) that give you access to arbitrary files and directories in the file system. Although other browsers are opposed to implementing these APIs because of safety and privacy concerns, this is not an issue for server-side runtimes.

I have heard that this API is not a great fit for server-side environments, but (as someone who hasn't looked in depth at the API and considered it) those concerns are not clear to me. So I opened this issue to have a place to discuss what it's missing, whether we can work with WHATWG to fix that, and also to discuss WinterCG extensions similar to `showOpenFilePicker()`.

**Comment by ekwoka** [comment]
I agree that I'm not too sure what the disagreement would be for the server context (outside of runtimes that inherently don't expose any FS as is).

Everything has unified around the node:fs system ,but the File System API seems to be quite a lot better thought out.

**Comment by Qard** [comment]
It lacks the full feature set and flexibility of the posix fs model, for example file encodings and permissions. Because the API is designed as a tree structure with methods on each element, such additional capabilities can not be added to it without violating the spec.

Additionally, the use of Web Streams and subsequently EventTarget makes the performance insufficient for Node.js at present.

It may be reasonable to have it as a secondary interface, but not really usable as the _only_ way to read and write files.

**Comment by andreubotella** [comment]
> Additionally, the use of Web Streams and subsequently EventTarget makes the performance insufficient for Node.js at present.

None of the FS APIs inherit from `EventTarget`. The only thing remotely connected is `FileReader`, and we're already not including that in the common minimum API even though we do include `File`.

As for streams, my understanding is that it is possible to achieve implementations of web streams with enough performance, so this would only be a temporary blocker for Node.js, right? After all, the usage of web APIs in Node.js is only ever going to increase, and web APIs won't be adopting Node.js's streams, so the pressure on Node.js's implementation to be more performant is only going to increase.

**Comment by Qard** [comment]
Web streams are based on EventTarget, so until the performance of _that_ is resolved it's unlikely we'd be able to get at all comparable performance with the existing fs API.

As I said though, the API simply lacks the capability of the existing API though, and has quite a lot of complexity in its object model, so I don't expect to see a Node.js implementation of it any time soon.

**Comment by andreubotella** [comment]
> Web streams are based on EventTarget, so until the performance of _that_ is resolved it's unlikely we'd be able to get at all comparable performance with the existing fs API.

I don't think that's correct. The streams spec has no mention of `EventTarget`, and the only times events are mentioned in normative text is related to passing stream chunks through a worker boundary when transferring streams, which Node.js would do with `EventEmitter`.

> As I said though, the API simply lacks the capability of the existing API though, and has quite a lot of complexity in its object model, so I don't expect to see a Node.js implementation of it any time soon.

Part of the point of this thread is to have a concrete list of capabilities that that API is missing, to see if we can work together with the WHATWG to make that possible.

**Comment by mk-pmb** [comment]
From a few minutes of reading MDN's OPFS docs and a quick scan of the Node.js fs API ToC, my impression is that polyfills will not be feasible for these Node.js features, unless you implement a full file system on top of OPFS:

<details><summary>[Toggle list]</summary>

```text
filehandle.chmod(mode)
filehandle.chown(uid, gid)
filehandle.stat([options])
filehandle.utimes(atime, mtime)
fsPromises.access(path[, mode])
fsPromises.chmod(path, mode)
fsPromises.chown(path, uid, gid)
fsPromises.lchmod(path, mode)
fsPromises.lchown(path, uid, gid)
fsPromises.lutimes(path, atime, mtime)
fsPromises.link(existingPath, newPath)
fsPromises.lstat(path[, options])
fsPromises.mkdir(path[, options])
fsPromises.mkdtemp(prefix[, options])
fsPromises.readlink(path[, options])
fsPromises.realpath(path[, options])
fsPromises.rename(oldPath, newPath)
fsPromises.stat(path[, options])
fsPromises.symlink(target, path[, type])
fsPromises.utimes(path, atime, mtime)
fsPromises.watch(filename[, options])
```

</details>

**Comment by jimmywarting** [comment]
The best thing we have so far is the low level: 

- `node:fs#openAsBlob`<br><sup>(which i think should have lived on the `node:fs/promise#openAsBlob` - but beside the point.</sup>

- Bun.js inherently badly written Blob / File / BunFile class you get from calling `Bun.file(path)`
  - https://github.com/denoland/deno/issues/27664#issuecomment-2888726897
  - https://github.com/oven-sh/bun/issues/5980

- This Deno [hack](https://github.com/denoland/deno/issues/27664#issuecomment-2888726897) for getting Blob/Files backed up by the file system.

this ☝️ are the core building blocks for creating something like: `await fileHandle.getFile()`

---

Really wished we could get something like a core `File System Access` for the backend, that works the same everywhere.

> I have heard that this API is not a great fit for server-side environments

I think the highest thing that got the most backslash was the edit-in-place for large files

> See [WICG/file-system-access issue #67](https://github.com/WICG/file-system-access/issues/67) for discussion around and desire for a "inPlace" mode for createWritable (where changes will be written to the actual underlying file as they are written to the writer, for example to support in-place modification of large files or things like databases). This is not currently implemented in Chrome. Implementing this is currently blocked on figuring out how to combine the desire to run malware checks with the desire to let websites make fast in-place modifications to existing large files. In-place writes are available for files in a [bucket file system](https://fs.spec.whatwg.org/#origin-private-file-system) via the [FileSystemSyncAccessHandle](https://fs.spec.whatwg.org/#filesystemsyncaccesshandle) interface.

i suppose servers don't need to have the same concern and it can allow edit-in-place

**Comment by jasnell** [comment]
I've recently started an implementation for Cloudflare workers. Mainly feeling it out and seeing what issues there are for server environments. There are definitely a few warts and places where some things don't line up but for the most part it's pretty straightforward. Later, once this impl is done I'll probably do one for node.js

**Comment by Ethan-Arrowood** [comment]
I'd be very interested in trying to contribute FS spec for Node.

**Comment by mb21** [comment]
After having shipped `Deno.readTextFile` and `Bun.file(path).text()` respectively, Deno and Bun and Cloudflare seem to have settled on `node:fs` compatibility for now. But even the most commonly used function – `node:fs/promises.readFile` – has the unergonomic `‌{ encoding: 'utf8' }` parameter. And a lot of the API is just a thin wrapper around POSIX file systems, which doesn’t seem well-suited for a virtual in-process file-system like in the browser. As mentioned [above](https://github.com/WinterTC55/proposal-minimum-common-api/issues/73#issuecomment-2204728010), the [OPFS](https://developer.mozilla.org/en-US/docs/Web/API/File_System_API/Origin_private_file_system) API is unfortunately also quite complex, and its handles-first API design is more verbose for simpler cases (this can be seen as either a feature or a bug). It seems hard to find a middle-ground with complete feature coverage.

But perhaps there is appetite to standardize a smaller number of simpler functions like:

- `readTextFile: (path: string) => Promise<string>`
- `readFile: (path: string) => Promise<Uint8Array>`

and perhaps a way to list files (similar to `fs.readDir` or `fs.glob`). Writing files and streaming is a whole other can of worms though.

## Prior art

### Val Town

- [readFile](https://docs.val.town/reference/file-io/#readfile)
- [listFiles](https://docs.val.town/reference/file-io/#listfiles)
- [listFilesByPath](https://docs.val.town/reference/file-io/#listfilesbypath)

### [Mastro](https://mastrojs.github.io/)
(which I’m working on)
- [readFile](https://jsr.io/@mastrojs/mastro/doc/~/readFile)
- [readTextFile](https://jsr.io/@mastrojs/mastro/doc/~/readTextFile)
- [findFiles](https://jsr.io/@mastrojs/mastro/doc/~/findFiles)
- [readDir](https://jsr.io/@mastrojs/mastro/doc/~/readDir)

They either use `node:fs` under the hood, or when we’re in the [Mastro VSCode extension](https://github.dev/mastrojs/template-basic), we call out to:

### VS Code extensions
that can run in the browser (like [github.dev](https://github.dev/mastrojs/template-basic)) can use its [Virtual FileSystem API](https://code.visualstudio.com/api/references/vscode-api#FileSystem) with methods like: 
- `vscode.workspace.fs.readFile`
- `vscode.workspace.fs.readDirectory` 
- `vscode.workspace.findFiles(pattern)`

P.S. There is also [node:vfs](https://blog.platformatic.dev/why-nodejs-needs-a-virtual-file-system)

**Comment by mk-pmb** [comment]
Starting with a small subset looks like a good step forward.

**Comment by mb21** [comment]
Another idea would be to model the JavaScript API after the [WebAssembly System Interface](https://github.com/WebAssembly/WASI/) (WASI) [filesystem proposal](https://github.com/WebAssembly/WASI/tree/main/proposals/filesystem/wit-0.3.0-draft).
